docs/language/learn-ql/ql-etudes/river-crossing.rst

```diff
@@ -0,0 +1,259 @@
+
+One way to restrict our paths to a finite number of river crossings is to define a 
+`member predicate <https://help.semmle.com/QL/ql-handbook/types.html#member-predicates>`__
+``State reachesVia(string path, int steps)``.
+The result of this predicate is any state that is reachable from the current state (``this``) via
+the given path in a specified finite number of steps.
+
+You can write this as a `recursive predicate <https://help.semmle.com/QL/ql-handbook/recursion.html>`__,
+with the following base case and recursion step:
+
+  - If ``this`` *is* the result state, then it (trivially) reaches the result state via an
+    empty path in zero steps.
+  - Any other state is reachable if ``this`` can reach an intermediate state (for some value of
+    ``path`` and ``steps``), and there is a ``safeFerry`` action from that intermediate
+    state to the result state.
+
+To ensure that the predicate is finite, you should restrict ``steps`` to a particular value,
+for example ``steps <= 7``.
+
+.. container:: toggle
+
+   .. container:: name
+
+      *Show/hide code*
+
+   .. literalinclude:: river-crossing-1.ql
+      :lines: 70-86
+
+However, although this ensures that the solution is finite, it can still contain loops if the upper bound
+for ``steps`` is large. In other words, you could get an inefficient solution by revisiting the same state
+multiple times.
+
+Instead of picking an arbitrary upper bound for the number of steps, you can avoid
+counting steps altogether. If you keep track of states that have already been visited and ensure
+that each ferrying action leads to a new state, the solution certainly won't contain any loops.
+
+To do this, change the member predicate to ``State reachesVia(string path, string visitedStates)``.
+The result of this predicate is any state that is reachable from the current state (``this``) via
+the given path without revisiting any previously visited states.
+
+  - As before, if ``this`` *is* the result state, then it (trivially) reaches the result state via an
+    empty path and an empty string of visited states.
+  - Any other state is reachable if  ``this`` can reach an intermediate state via some path, without
+    revisiting any previous states, and there is a ``safeFerry`` action from the intermediate state to
+    the result state.
+    (Hint: To check whether a state has previously been visited, you could check if
+    there is an `index of <https://help.semmle.com/QL/ql-spec/language.html#built-ins-for-string>`__
+    ``visitedStates`` at which the state occurs.)
+
+.. container:: toggle
+
+   .. container:: name
+
+      *Show/hide code*
+
+   .. literalinclude:: river-crossing.ql
+      :lines: 83-102
+
+Display the results
+~~~~~~~~~~~~~~~~~~~
+
+Once you've defined all the necessary classes and predicates, write a `select clause <https://help.semmle.com/QL/ql-handbook/queries.html#select-clauses>`__
+that returns the resulting path.
+
+.. container:: toggle
```

Comment:
I think it would be good to add a sentence explaining the use of `_` in the select clause (including a link to 
https://help.semmle.com/QL/ql-handbook/expressions.html#don-t-care-expressions).